drivers/sensor/si1153/si1153.c

```diff
@@ -0,0 +1,655 @@
+	if (samples->ch2 & 0x800000) {
+		samples->ch2 |= 0xFF000000;
+	}
+
+	samples->ch3  = buffer[10] << 16;
+	samples->ch3 |= buffer[11] <<  8;
+	samples->ch3 |= buffer[12];
+	if (samples->ch3 & 0x800000) {
+		samples->ch3 |= 0xFF000000;
+	}
+
+}
+
+static int si1153_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+#ifndef CONFIG_SI1153_OWN_THREAD
+	struct si1153_data *data = dev->driver_data;
+	struct si115x_samples samples;
+
+	si115x_force(data);
+
+	si115x_handler(data, &samples);
+
+	data->gesture = si115x_gesture_algorithm(&samples);
+	data->ch0 = samples.ch0;
+	data->ch1 = samples.ch1;
+	data->ch2 = samples.ch2;
+	data->ch3 = samples.ch3;
+
+#ifdef SI1153_DEBUG
+	char *strgesture[] = { "NONE", "UP", "DOWN", "LEFT", "RIGHT", "PROX" };
+
+	SYS_LOG_DBG("Gesture %d %s %d %d %d %d\n",
+	       data->gesture, strgesture[data->gesture],
+	       data->ch0,
+	       data->ch1,
+	       data->ch2,
+	       data->ch3);
+#endif
+
+#endif
+	return 0;
+}
+
+static int si1153_channel_get(struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct si1153_data *data = dev->driver_data;
+
+	switch (chan) {
+
+	case SENSOR_CHAN_LIGHT:
+	case SENSOR_CHAN_IR:
+		val->val1 = data->ch0;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_PROX:
+		val->val1 = data->ch1;
+		val->val2 = 0;
+		val++;
+		val->val1 = data->ch2;
+		val->val2 = 0;
+		val++;
+		val->val1 = data->ch3;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_GESTURE:
+		val->val1 = data->gesture;
+		val->val2 = 0;
+		break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static const struct sensor_driver_api si1153_api_funcs = {
+	.sample_fetch  = si1153_sample_fetch,
+	.channel_get   = si1153_channel_get,
+};
+
+#ifdef CONFIG_SI1153_OWN_THREAD
+static void si1153_thread(void *p1, void *p2, void *p3)
+{
+	struct si1153_data *data = (struct si1153_data *)p1;
+	struct si115x_samples samples;
+	int old = -1;
+
+	while (1) {
+
+		si115x_force(data);
+
+		si115x_handler(data, &samples);
+
+		data->gesture = si115x_gesture_algorithm(&samples);
+
+		data->ch0 = samples.ch0;
+		data->ch1 = samples.ch1;
+		data->ch2 = samples.ch2;
+		data->ch3 = samples.ch3;
+#ifdef SI1153_DEBUG
+		SYS_LOG_INFO("Gesture %d %d %d %d\n",
+		       data->ch0,
+		       data->ch1,
+		       data->ch2,
+		       data->ch3);
+#endif
+		if (data->gesture != old) {
+#ifdef SI1153_DEBUG
+			char *strgesture[] = { "NONE", "UP", "DOWN", "LEFT", "RIGHT", "PROX" };
+
+			SYS_LOG_DBG("Gesture %d %s\n", data->gesture, strgesture[data->gesture]);
+#endif
+			old = data->gesture;
+		}
+
+		k_busy_wait(500);
+	}
+}
+#endif
+
+
+static int si1153_init(struct device *dev)
+{
+	struct si1153_data *data = dev->driver_data;
+
+	data->i2c_master = device_get_binding(CONFIG_SI1153_I2C_DEV_NAME);
+	if (!data->i2c_master) {
+		SYS_LOG_DBG("i2c master not found: %s",
+			    CONFIG_SI1153_I2C_DEV_NAME);
+		return -EINVAL;
+	}
+
+	data->i2c_slave_addr = CONFIG_SI1153_I2C_DEV_ADDRESS;
```

Comment:
Do you need this in the data structure? You can probably put in the config structure instead and save a little ram.